src/Raven.Server/Documents/Indexes/Persistence/Corax/CoraxIndexFacetedReadOperation.cs

```diff
@@ -0,0 +1,428 @@
+
+        var parameters = new CoraxQueryBuilder.Parameters(_indexSearcher, null, null, query, _index, query.QueryParameters, _queryBuilderFactories,
+            _fieldMappings, null, null, -1, null);
+        var baseQuery = CoraxQueryBuilder.BuildQuery(parameters, out var isBinary);
+        var coraxPageSize = CoraxGetPageSize(_indexSearcher, facetQuery.Query.PageSize, query, isBinary);
+        var ids = CoraxIndexReadOperation.QueryPool.Rent(coraxPageSize);
+        using var analyzersScope = new AnalyzersScope(_indexSearcher, _fieldMappings, _index.Definition.HasDynamicFields);
+        int read = 0;
+        while ((read = baseQuery.Fill(ids)) != 0)
+        {
+            for (int docId = 0; docId < read; docId++)
+            {
+                var entryReader = _indexSearcher.GetReaderFor(ids[docId]);
+                foreach (var result in results)
+                {
+                    token.ThrowIfCancellationRequested();
+
+                    using var facetTiming = queryTimings?.For($"{nameof(QueryTimingsScope.Names.AggregateBy)}/{result.Key}");
+
+                    if (result.Value.Ranges == null || result.Value.Ranges.Count == 0)
+                    {
+                        facetsByName ??= new Dictionary<string, Dictionary<string, FacetValues>>();
+
+                        HandleFacetsPerDocument(ref entryReader, result, facetsByName, facetQuery.Legacy, facetTiming, analyzersScope, token);
+                        continue;
+                    }
+
+                    // Cache facetByRange because we will fulfill data in batches instead of whole collection
+                    facetsByRange ??= new();
+                    if (facetsByRange.TryGetValue(result.Key, out var facetValues) == false)
+                    {
+                        facetValues = new();
+                        facetsByRange.Add(result.Key, facetValues);
+                    }
+
+                    HandleRangeFacetsPerDocument(ref entryReader, result, facetQuery.Legacy, facetTiming, facetValues, token);
+                }
+            }
+
+            token.ThrowIfCancellationRequested();
+        }
+
+        UpdateRangeResults();
+        
+        UpdateFacetResults(results, query, facetsByName);
+
+        CompleteFacetCalculationsStage(results);
+        CoraxIndexReadOperation.QueryPool.Return(ids);
+        return results.Values
+            .Select(x => x.Result)
+            .ToList();
+
+        void UpdateRangeResults()
+        {
+            foreach (var result in results)
+            {
+                if (facetsByRange != null)
+                    foreach (var kvp in facetsByRange)
+                    {
+                        foreach (var inner in kvp.Value)
+                        {
+                            if (inner.Value.Any == false)
+                                continue;
+
+                            result.Value.Result.Values.AddRange(inner.Value.GetAll());
```

Comment:
I don't understand what you are doing here. 
Shouldn't you be copying the ranges?